### PR TITLE
Don't use $rate as an object (it is an array)

### PR DESCRIPTION
**Pull request by flancer64**
### Description (*)
`$rate` is an array and it does not contain `id` item. We can use rate's `code` item as part of `id` for HTML nodes.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by Stepa4man** [comment]
@magento run all tests

File: app/code/Magento/Sales/view/adminhtml/templates/order/totals/tax.phtml [hunk 1]
```diff
@@ -59,7 +59,11 @@ $randomHelper = $block->getData('randomHelper');
             ?>
 
             <?php foreach ($rates as $rate): ?>
-                <tr id="rate-<?= /* @noEscape */ $rate->getId() ?>"
+                <?php
+                    $rateCode = strtolower($rate['code']);
+                    $rateId = str_replace(' ', '-', $rateCode);
+                ?>
```
Comment (Stepa4man) [inline app/code/Magento/Sales/view/adminhtml/templates/order/totals/tax.phtml]: Please, don't hold logic in the template. Can you put this to view Model?

File: app/code/Magento/Sales/view/adminhtml/templates/order/totals/tax.phtml [hunk 2]
```diff
@@ -59,7 +59,11 @@ $randomHelper = $block->getData('randomHelper');
...
+                <tr id="rate-<?= /* @noEscape */ $rateId ?>"
```
Comment (Stepa4man) [inline app/code/Magento/Sales/view/adminhtml/templates/order/totals/tax.phtml]: Why should we keep **$rateId** as the name of the variable?

**Review by Stepa4man — CHANGES_REQUESTED** [review]
Hi @flancer64 
Thanks for your contribution

Could you, please take a look at my comments?

**Comment by flancer64** [comment]
Hello, @Stepa4man .

Sorry, I cannot "_move logic from template to view Model_". I don't know what a _right_ logic should be here. I've just fixed the bug. Someone  wrote this:
```
$rate->getId()
```

for array `$rate`. It interrupts templates rendering. 

I'm sorry again, but I have no answer to question "_Why should we keep $rateId as the name of the variable?_" :( You shouldn't. It just my own choice.

Please, feel free to reject this pull request if it doesn't match your rules.

**Comment by engcom-Golf** [comment]
Hi, @flancer64, are you able to continue with this pull request?
As a unique value, you can use tax_id field and if it is still necessary you can move logic to view model and use in this template.
Instructions can be found by the following link [View models](https://devdocs.magento.com/guides/v2.4/extension-dev-guide/view-models.html)

Also could you please cover your code changes with an MFTF test? You can use the following [Magento Functional Testing Framework Guide](https://devdocs.magento.com/mftf/docs/introduction.html) to get acquainted with it.
If you have any additional questions or appointments feel free to ask, I will try to help,

Thank you!

**Comment by flancer64** [comment]
@engcom-Golf , I cannot use `tax_id`. There was no such field under debugger in the structure. I don't know why this field was omitted on rate data selection from DB but it was:

image

I agree that the better way is to use `tax_id` field here instead of `code` but we need to fix code that selects data from DB first. I'm sorry, but It's too deep for me, I don't have so much free time to check side effects. I've created this "dirty fix" and it's pretty enough for me and my projects.

**Comment by flancer64** [comment]
Maybe we can use `position` field or even 'foreach' loop index here.

**Comment by engcom-Golf** [comment]
@flancer64, please look at \Magento\Tax\Model\ResourceModel\Calculation::getCalculationProcess method. There should be an ability to add rate id to the output array.

**Comment by flancer64** [comment]
Thanks, @engcom-Golf . This is the way.

This SQL is composed to select rates data in `\Magento\Sales\Block\Adminhtml\Order\Totals\Tax::getFullTaxInfo`:
```sql
SELECT
	`main_table` .*
FROM
	`sales_order_tax` AS `main_table`
WHERE
	(main_table.order_id = 1234)
ORDER BY
	`process` ASC
``` 
so, we can use `sales_order_tax.tax_id` as identifier for tax items.

**Comment by engcom-Golf** [comment]
@magento run all tests

**Comment by engcom-Golf** [comment]
Hi, @flancer64, thank you for your changes, but as we can see from failed tests, this approach is not acceptable. I have made some investigations and found that rates are taken not only from the database. So tax_id can be not present in rates array. Suppose we can use 'for each' loop index, as you have proposed before.

**Comment by flancer64** [comment]
> Suppose we can use 'for each' loop index, as you have proposed before.

@engcom-Golf , done.

**Comment by engcom-Golf** [comment]
@magento run all tests

**Review by Stepa4man — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @Stepa4man, thank you for the review.
[ENGCOM-8353](https://jira.corp.magento.com/browse/ENGCOM-8353) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
Seems like this PR fixes https://github.com/magento/magento2/issues/30094.

We also have another PR https://github.com/magento/magento2/pull/30516 that fixes this issue.

**Review by engcom-Delta — CHANGES_REQUESTED** [review]
@flancer64 sorry for delay
Due to [Magento Definition of Done](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing_dod.html) all code must be covered by tests. 

Could you add automation test that cover your changes?

**Comment by flancer64** [comment]
@engcom-Delta , I'm sorry, I couldn't.

**Comment by ihor-sviziev** [comment]
Hi @flancer64,

I just analyzed another solution that we have - seems like array index solution potentially might cause ID duplication, as in some other place on the page might be same id `rate-1`, or `rate-0`. I believe using random string is better and this approach is used in https://github.com/magento/magento2/pull/30516.

I'm closing this PR as we just have better solution for the same issue in https://github.com/magento/magento2/pull/30516.

Thank you so much for your contribution!

**Pull request closed without merging**